[MRG] Issue #7987: Embarrassingly parallel "n_restarts_optimizer" in GaussianProcessRegressor

amanp10 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
Fixes #7987  

It adds the Embarrassingly Parallel helper in GaussianProcessRegressor class.

I have used 'check_pickle=False' for the 'delayed' function as it was not working otherwise. However, I am not sure about the logic behind this.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

amanp10 committed:
Issue #7987

jnothman commented:
Please modify tests to use the new parameter and check results are identical

On 7 December 2016 at 20:32, Aman Pratik <notifications@github.com> wrote:

> Fixes #7987 <https://github.com/scikit-learn/scikit-learn/issues/7987>
>
> It adds the Embarrassingly Parallel helper in GaussianProcessRegressor
> class.
>
> I have used 'check_pickle=False' for the 'delayed' function as it was not
> working otherwise. However, I am not sure about the logic behind this.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scikit-learn/scikit-learn/pull/7997
> Commit Summary
>
>    - Issue #7987
>
> File Changes
>
>    - *M* sklearn/gaussian_process/gpr.py
>    <https://github.com/scikit-learn/scikit-learn/pull/7997/files#diff-0>
>    (17)
>
> Patch Links:
>
>    - https://github.com/scikit-learn/scikit-learn/pull/7997.patch
>    - https://github.com/scikit-learn/scikit-learn/pull/7997.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-lp-9_ddIZjNdPilRR9XAEkL2YEks5rFn0ngaJpZM4LGX5B>
> .
>

lesteve commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -209,12 +217,17 @@ def obj_func(theta, eval_gradient=True):
                         self._constrained_optimization(obj_func, theta_initial,
                                                        bounds))
+                Parallel(n_jobs=self.n_jobs)(delayed(optima_iterations,
+                                                     check_pickle=False)()
```
I would be really surprised if using an inner function like this was working even with `check_pickle=False` for `n_jobs != 1`... You will need to move optima_iterations to the module scope (or class-level maybe).

Drawing random numbers in parallel processes seems a bit dangerous too. And I don't think you will be able to match the result of `n_jobs=1` with `n_jobs!=-1` this way. An alternative would be to draw all the random numbers first and then pass them to the function inside Parallel.

Optional: refactoring optima_iterations into a pure function that does not take self as argument is probably a good idea to reduce surprising behaviours.

lesteve replied:
To illustrate what I mean:

```py
from joblib import Parallel, delayed

import numpy as np


def func(rng):
    return rng.randn(2)

rng = np.random.RandomState(0)

n_jobs=2
result = Parallel(n_jobs=n_jobs)(delayed(func)(rng) for x in [1, 2, 3])
print(result)
```

with n_jobs=1:
```
[array([ 1.76405235,  0.40015721]), array([ 0.97873798,  2.2408932 ]), array([ 1.86755799, -0.97727788])]
```

with n_jobs=2
```
[array([ 1.76405235,  0.40015721]), array([ 1.76405235,  0.40015721]), array([ 1.76405235,  0.40015721])]
```

What happens in the n_jobs=2 case is that each subprocess gets a copy of rng so the random numbers are the same for each value of x.

amanp10 committed:
tests_gpr

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 committed:
Issue #7987 2

amanp10 commented:
I have made the updates suggested by @jnothman . Please have a look.
Also, according to @lesteve I have changed the scope of optima_iterations to the Class and also taken care of the random numbers issue. Please have a look.
After your approval I will change the Title to [MRG] from [WIP]

jnothman commented:
MRG means "ready for full review before merging" so you should change that now.

lesteve commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -131,14 +137,21 @@ def optimizer(obj_func, initial_theta, bounds):
+        self.n_jobs = n_jobs
+
+    def _optima_iterations(self, optima, bounds, theta_initial, obj_func):
+        optima.append(
```
How can that work ? If you read my previous https://github.com/scikit-learn/scikit-learn/pull/7997#discussion_r91288573, when n_jobs!=1, each worker gets a copy of `optima` and  this should not change `optima` in the main process ...

You should return `self._constrained_optimizations(...)` and do the `optima.append` in `fit`.

lesteve commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -209,12 +222,17 @@ def obj_func(theta, eval_gradient=True):
-                                                       bounds))
+                    theta_initials.append(theta_initial)
+                Parallel(n_jobs=self.n_jobs)(delayed(self._optima_iterations,
+                                                     check_pickle=False)
```
Is check_pickle=False needed?

amanp10 replied:
It gives an error "TypeError: can't pickle instancemethod objects" when check_pickle=True.
Hence I did otherwise. Please let me know if it is wrong.

amanp10 commented:
I will look into these issues and get back to you asap.

lesteve commented:
> I will look into these issues and get back to you asap.

I forgot to say, this is slightly worrying that the test was passing ... it is probably too forgiving.

amanp10 committed:
Issue #7987 3

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 commented:
@lesteve I have made the changes you suggested, however some tests are failing (I have mentioned them in the test file). Please go through the changes and give your reviews. Also, I have made few changes in 2 example files to test this new parameter. Thank you for devoting so much of your time.

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -107,6 +108,11 @@ def optimizer(obj_func, initial_theta, bounds):
         number generator.
 
+    n_jobs : int, default: 1
+        n_jobs is the number of workers requested by the callers.
```
Please copy the docstring from within scikit-learn, rather than from joblib. This text is a bit obscure.

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -209,12 +222,17 @@ def obj_func(theta, eval_gradient=True):
-                        self._constrained_optimization(obj_func, theta_initial,
-                                                       bounds))
+                    theta_initials.append(theta_initial)
+                Parallel(n_jobs=self.n_jobs)(delayed(self._optima_iterations,
```
there are a few ways you could make this line wrapping more readable! please find one...

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -209,12 +220,16 @@ def obj_func(theta, eval_gradient=True):
                 bounds = self.kernel_.bounds
-                for iteration in range(self.n_restarts_optimizer):
+                theta_initials = []
+                for i in range(self.n_restarts_optimizer):
```
use the size parameter of `rng.uniform` to generate n_restarts_optimizer elements in one fast operation

amanp10 committed:
Issue #7987 4

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 committed:
Issue #7987 5

amanp10 commented:
I have made the changes @jnothman suggested above.

amanp10 commented:
Hello,
I am not able to understand why some of the tests are failing. Any kind of help is welcome.

def test_n_jobs_parallel():
    # Test to check the functioning of n_jobs parameter.
    for kernel in kernels:
        gpr1 = GaussianProcessRegressor(kernel=kernel, n_jobs=1,
                                        n_restarts_optimizer=5).fit(X, y)
        gpr2 = GaussianProcessRegressor(kernel=kernel, n_jobs=2,
                                        n_restarts_optimizer=5).fit(X, y)
        gpr3 = GaussianProcessRegressor(kernel=kernel, n_jobs=-1,
                                        n_restarts_optimizer=5).fit(X, y)
        y1, y1_cov = gpr1.predict(X, return_cov=True)
        y2, y2_cov = gpr2.predict(X, return_cov=True)
        y3, y3_cov = gpr3.predict(X, return_cov=True)
        # Successfully passed tests
        assert_almost_equal(y1, y2)
        assert_almost_equal(y1, y3)
        assert_almost_equal(y1_cov, y2_cov)
        assert_almost_equal(y1_cov, y3_cov)
        # Failing tests
        assert_almost_equal(gpr1.alpha_, gpr2.alpha_)
        assert_almost_equal(gpr1.alpha_, gpr3.alpha_)
        assert_almost_equal(gpr1.log_marginal_likelihood_value_,
                            gpr2.log_marginal_likelihood_value_)
        assert_almost_equal(gpr1.log_marginal_likelihood_value_,
                            gpr3.log_marginal_likelihood_value_)

I also tried decreasing the precision value (required decimal places). This problem is arising for few kernels only.

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 committed:
Issue#7987 6

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 committed:
Issue#7897 7

jnothman commented:
(Meaningful commit messages would help us know what to expect when we review changes.)

amanp10 commented:
@jnothman I will definitely take care of it next time. For now,
I have added the usage of Embarrassingly parallel helper in 'n_restarts_optimizer'  iterations for better optima. You can see this in the 'sklearn/gaussian_process/gpr.py' file.
Then I added a new test in 'test_gpr.py' to test the functioning of this parallel code. However some of the assertions failed, as I have commented in the 'test_gpr.py' file.
I also made a slight change in the 'examples/gaussian_process/plot_gpr_noisy.py & plot_gpr_noisy_targets.py' to test the parallel code by using different n_jobs parameter value.

amanp10 committed:
Issue#7987 Failing tests corrected

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 commented:
@jnothman I have removed the failing tests since those test comparisons were failing even for the same n_jobs parameter. This is probably because of the random seeding leading to different 'alpha_' and 'log_marginal_likelihood_value_' irrespective of n_jobs parameter.

jnothman commented:
you should be able to use random_state=42 to ensure determinism, and we'd
like that determinism to be maintained even with different n_jobs, if
possible.

On 14 December 2016 at 16:51, Aman Pratik <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> I have removed the failing tests
> since those test comparisons were failing even for the same n_jobs
> parameter. This is probably because of the random seeding leading to
> different 'alpha_' and 'log_marginal_likelihood_value_' irrespective of
> n_jobs parameter.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997#issuecomment-266949056>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65JJ1QeW74hDS3ueJ4vQCwYEd_Xiks5rH4PUgaJpZM4LGX5B>
> .
>

jnothman commented on sklearn/gaussian_process/tests/test_gpr.py:
```diff
@@ -314,3 +314,29 @@ def test_duplicate_input():
+def test_n_jobs_parallel():
+    # Test to check the functioning of n_jobs parameter.
+    for kernel in kernels:
+        gpr1 = GaussianProcessRegressor(kernel=kernel, n_jobs=1,
```
include `random_state=<fixed int>` for each of these, and then try the tests again

amanp10 committed:
Issue#7987 All tests passed, for random_state=42

amanp10 commented:
@jnothman  In the last commit I used random_state=42 and all the tests have passed. Please have a look.

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -209,12 +219,22 @@ def obj_func(theta, eval_gradient=True):
+                                     len(self.kernel_.theta)))
+                # 'n_restarts_optimizer' iterations using Embarrassingly
+                # Parallel helper
+                delayed_optima = delayed(self._optima_iterations,
```
why can't you use `_constrained_optimization` here directly?

amanp10 replied:
should have thought that myself.

jnothman commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -107,6 +108,10 @@ def optimizer(obj_func, initial_theta, bounds):
 
+    n_jobs : int, default: 1
+        Number of CPU cores used during the 'n_restarts_optimizer' iterations.
+        If given a value of -1, all cores are used.
```
Please note that `optimizer`, if provided, must be thread-safe.

amanp10 replied:
I will look into it.

amanp10 replied:
I will add an entry. I did not get LGTM???

amanp10 replied:
Regarding the `optimizer`, should I add a warning in the docstring or do something else?

jnothman reviewed (commented):
otherwise, this LGTM!

While awaiting another review, please add an entry into whats_new.rst

jnothman commented:
you're right, @lesteve, it'd be worth getting a quick benchmark of timing
multithreading against multiprocessing...

On 14 December 2016 at 20:27, Loïc Estève <notifications@github.com> wrote:

> *@lesteve* commented on this pull request.
> ------------------------------
>
> In sklearn/gaussian_process/gpr.py
> <https://github.com/scikit-learn/scikit-learn/pull/7997#pullrequestreview-12863087>
> :
>
> > -                for iteration in range(self.n_restarts_optimizer):
> -                    theta_initial = \
> -                        self.rng.uniform(bounds[:, 0], bounds[:, 1])
> -                    optima.append(
> -                        self._constrained_optimization(obj_func, theta_initial,
> -                                                       bounds))
> +                # Initial theta values for 'n_restarts_optimizer' iterations
> +                # for better optima
> +                theta_initials = \
> +                    self.rng.uniform(bounds[:, 0], bounds[:, 1],
> +                                     size=(self.n_restarts_optimizer,
> +                                     len(self.kernel_.theta)))
> +                # 'n_restarts_optimizer' iterations using Embarrassingly
> +                # Parallel helper
> +                delayed_optima = delayed(self._optima_iterations,
> +                                         check_pickle=False)
>
> Still my comment about check_pickle=False. My preference would be to use
> pure functions in Parallel (maybe a staticmethod will do, this is a quick
> thing to try).
>
> Also just curious, do you need backend='threading'?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997#pullrequestreview-12863087>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67eGQIp_zSboImOglMEpbjwkHcNGks5rH7ZogaJpZM4LGX5B>
> .
>

lesteve commented:
> you're right, @lesteve, it'd be worth getting a quick benchmark of timing multithreading against multiprocessing...

and also against n_jobs=1 to gauge how much quicker this is.

lesteve commented on examples/gaussian_process/plot_gpr_noisy.py:
```diff
@@ -39,7 +39,8 @@
 gp = GaussianProcessRegressor(kernel=kernel,
-                              alpha=0.0).fit(X, y)
+                              alpha=0.0, n_jobs=2,
+                              n_restarts_optimizer=3).fit(X, y)
```
Is there a particular reason why you modified the examples?

Note: `n_jobs!=1` is not really recommended in examples because it will cause a problem in Windows if we do not use `if __name__ == '__main__'`.

lesteve replied:
General guideline: if there is not a strong reason behind a change, just keep it as it was before. The more focused the changes, the easier and more pleasant it is to review them.

amanp10 replied:
I will surely take care of it from now on.

amanp10 commented on sklearn/gaussian_process/gpr.py:
```diff
@@ -209,12 +219,22 @@ def obj_func(theta, eval_gradient=True):
+                # 'n_restarts_optimizer' iterations using Embarrassingly
+                # Parallel helper
+                delayed_optima = delayed(self._optima_iterations,
+                                         check_pickle=False)
```
I will try the static method. There was  some error when backend=default so I changed it to 'threading', though I will check it again.

lesteve replied:
Adding inline comments for non obvious changes like this can ease the review process. Of course it's better if you have a more detailed explanation rather than something like "it did not work for some reason, so I try something else and now it works, oh well"

amanp10 replied:
I understand, next time I will discuss it here before taking such a decision.

amanp10 commented:
According to @jnothman I used '_constrained_optimization' directly without using any other function. I also used staticmethod decorator for '_constrained_optimization'.
However, when check_pickle=True , I get the following error,
**TypeError: can't pickle function objects**
There is a workaround for this using extrenal libraries but I am not sure whether to use it here.

Also, when backend=default, I get the following error,
**pickle.PicklingError: Can't pickle <function _constrained_optimization at 0x7fdac5486230>: it's not found as sklearn.gaussian_process.gpr._constrained_optimization**
Using an external function instead of '_constrained_optimization' might solve the issue.
What should I do next?

amanp10 committed:
Issue#7987 examples as before, whats_new.rst changed

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 commented:
Should I add a benchmark test for different `n_jobs` parameter with backend='threading' for now?

jnothman commented:
Yes, benchmark with backend='threading'

jnothman commented:
But it's also extremely easy to pull `_constraint_optimization` out as a function of `(optimizer, obj_func, initial_theta, bounds)` such that it is importable/picklable. However, `obj_func` and `optimizer` remain unpicklable, so I think we're better off requiring that the threading backend only is used, and that `check_pickle=False`.

amanp10 committed:
Added Timing Benchmark Test(Failing)

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 commented:
I have added the Benchmark test but its failing. I have used python timeit, however it seems that with `n_jobs=1` the `fit` method executes faster than with `n_jobs=2` and `n_jobs=-1`.

jnothman commented:
Yes, that'll happen if there's not enough time spent in parts of the code
where the GIL is released. What shape of dataset are you doing that
benchmark over? I suspect that indeed, unless this were rewritten in
Cython, we'd not be releasing the GIL enough.

I have a feeling that @lesteve's request for a benchmark was well
justified, and we will find no gains here, unless we can avoid closures and
use multiprocessing.

On 28 December 2016 at 15:00, Aman Pratik <notifications@github.com> wrote:

> I have added the Benchmark test but its failing. I have used python
> timeit, however it seems that with n_jobs=1 the fit method executes
> faster than with n_jobs=2 and n_jobs=-1.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997#issuecomment-269422150>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65nxT2sEAZOufdRuV7uEP_rttvtVks5rMd7ygaJpZM4LGX5B>
> .
>

amanp10 commented:
The shapes were X->(6, 1) and y->(6,).
What do you suggest I do?

jnothman commented:
oh well, that is very small for a benchmark. benchmarks teens to use
something closer to a realistic but large sample size, since performance on
small datasets rarely matters.

On 28 Dec 2016 8:50 pm, "Aman Pratik" <notifications@github.com> wrote:

> The shapes were X->(6, 1) and y->(6,).
> What do you suggest I do?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997#issuecomment-269453423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60lw7UtP7iKeg4uWOAKqMd2ib3KDks5rMjDHgaJpZM4LGX5B>
> .
>

amanp10 commented:
What do you suggest I do next? I used the `Boston House` dataset but the tests were still failing. Should I try changing the function scopes so as to use `backend="multiprocessing"`?

jnothman commented:
Oh, I'd not realised you'd added tests for benchmarking. Usually we don't do this, but build a benchmarking script and upload it as a gist, or if we want to be able to repeat a benchmark over time, contribute it to the `benchmarks/` directory in scikit-learn.

While this means we only test one user's platform at a time, it means we can do things like flexibly plot multiple benchmarks.

jnothman commented:
I'd rather see plots showing the size of the effect rather than just a passed test (which may fail if the machine is having a bad day, which is no good for continuous integration)

amanp10 committed:
Benchmark test added in Benchmark/ folder

amanp10 committed:
Merge branch 'master' of https://github.com/amanp10/scikit-learn into mybranch

amanp10 commented:
I have added the Benchmark test in the `benchmarks/` folder. Please have a look.

jnothman commented:
I can only see n_jobs != 1 being an improvement on the 5th and to a lesser extent 6th kernels. (Though your measurements are not very robust as you make only one timing of each case.)

The gains so far seem to be quite marginal, and unless we can find a reasonable kernel or dataset size in which the gains are greater, perhaps this work is not of sufficient utility to merge :\

amanp10 commented:
@jnothman  What do you suggest me to do? Should I look for more kernels or datasizes? Or should I stop working on this issue?

jnothman commented:
I suspect we're not going to get anywhere with this. But if you have a
moment to try different sizes, etc, why not?

On 11 January 2017 at 14:47, Aman Pratik <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> What do you suggest me to do?
> Should I look for more kernels or datasizes? Or should I stop working on
> this issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997#issuecomment-271770768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63mZMail7TVE8olCBxPLSWwGM0XRks5rRFDmgaJpZM4LGX5B>
> .
>

amanp10 commented:
I tried using different datasizes but it was not of any use. I dont have much knowledge about kernels so I will not be working on this issue unless advised otherwise.

jnothman commented:
I think I'll close this, then.

On 12 January 2017 at 15:28, Aman Pratik <notifications@github.com> wrote:

> I tried using different datasizes but it was not of any use. I dont have
> much knowledge about kernels so I will not be working on this issue unless
> advised otherwise.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7997#issuecomment-272073348>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66SeS070X0N4vi4sfdfAVgzLvWCGks5rRawAgaJpZM4LGX5B>
> .
>

jnothman commented:
Sorry about the wasted effort. I hope you learnt from it, at least :|

On 12 January 2017 at 15:43, Joel Nothman <joel.nothman@gmail.com> wrote:

> I think I'll close this, then.
>
> On 12 January 2017 at 15:28, Aman Pratik <notifications@github.com> wrote:
>
>> I tried using different datasizes but it was not of any use. I dont have
>> much knowledge about kernels so I will not be working on this issue unless
>> advised otherwise.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/7997#issuecomment-272073348>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz66SeS070X0N4vi4sfdfAVgzLvWCGks5rRawAgaJpZM4LGX5B>
>> .
>>
>
>

amanp10 commented:
I did learn a lot. Thanks to you @jnothman  and @lesteve
